File: Data_Structures/Queue/Deque/C#/Deque.cs
/*
 * Deque - Double-ended queue backed by LinkedList
 *
 * Operations: PushFront/PushBack O(1), PopFront/PopBack O(1), Peek O(1)
 */

using System;
using System.Collections.Generic;

class Deque {
    private LinkedList<int> list;

    public Deque() {
        list = new LinkedList<int>();
    }

    public void PushFront(int value) { list.AddFirst(value); }
    public void PushBack(int value)  { list.AddLast(value); }

    public int PopFront() {
        if (IsEmpty()) throw new InvalidOperationException("Deque is empty");
        int val = list.First.Value;
        list.RemoveFirst();
        return val;
    }

    public int PopBack() {
        if (IsEmpty()) throw new InvalidOperationException("Deque is empty");
        int val = list.Last.Value;
        list.RemoveLast();
        return val;
    }

    public int PeekFront() {
        if (IsEmpty()) throw new InvalidOperationException("Deque is empty");
        return list.First.Value;
    }

    public int PeekBack() {
        if (IsEmpty()) throw new InvalidOperationException("Deque is empty");
        return list.Last.Value;
    }

    public bool IsEmpty() => list.Count == 0;
    public int  Size()    => list.Count;

    static void Main() {
        Deque d = new Deque();

        d.PushBack(1);
        d.PushFront(0);
        d.PushBack(2);

        Console.WriteLine($"Front: {d.PeekFront()}, Back: {d.PeekBack()}");
        Console.WriteLine($"Pop front: {d.PopFront()}");
        Console.WriteLine($"Pop back: {d.PopBack()}");
        Console.WriteLine($"Remaining front: {d.PeekFront()}");
    }
}

File: Data_Structures/Queue/Deque/Java/Deque.java
/*
 * Deque - Double-ended queue backed by a doubly-linked list
 *
 * Operations: pushFront/pushBack O(1), popFront/popBack O(1), peek O(1)
 */

public class Deque {
    private java.util.LinkedList<Integer> list;

    public Deque() {
        list = new java.util.LinkedList<>();
    }

    public void pushFront(int value) { list.addFirst(value); }
    public void pushBack(int value)  { list.addLast(value); }

    public int popFront() {
        if (isEmpty()) throw new RuntimeException("Deque is empty");
        return list.removeFirst();
    }

    public int popBack() {
        if (isEmpty()) throw new RuntimeException("Deque is empty");
        return list.removeLast();
    }

    public int peekFront() {
        if (isEmpty()) throw new RuntimeException("Deque is empty");
        return list.getFirst();
    }

    public int peekBack() {
        if (isEmpty()) throw new RuntimeException("Deque is empty");
        return list.getLast();
    }

    public boolean isEmpty() { return list.isEmpty(); }
    public int size()        { return list.size(); }

    public static void main(String[] args) {
        Deque d = new Deque();

        d.pushBack(1);
        d.pushFront(0);
        d.pushBack(2);

        System.out.println("Front: " + d.peekFront() + ", Back: " + d.peekBack());
        System.out.println("Pop front: " + d.popFront());
        System.out.println("Pop back: " + d.popBack());
        System.out.println("Remaining front: " + d.peekFront());
    }
}
